libbeat/cmd/instance/locks/lock.go

```diff
@@ -0,0 +1,241 @@
+	"github.com/elastic/elastic-agent-system-metrics/metric/system/resolve"
+)
+
+type Locker struct {
+	fileLock  *flock.Flock
+	logger    *logp.Logger
+	beatName  string
+	filePath  string
+	beatStart time.Time
+}
+
+type pidfile struct {
+	Pid       int       `json:"pid"`
+	WriteTime time.Time `json:"write_time"`
+}
+
+var (
+	// ErrAlreadyLocked is returned when a lock on the data path is attempted but
+	// unsuccessful because another Beat instance already has the lock on the same
+	// data path.
+	ErrAlreadyLocked = fmt.Errorf("data path already locked by another beat. Please make sure that multiple beats are not sharing the same data path (path.data).")
+)
+
+// New returns a new pid-aware file locker
+// all logic, including checking for existing locks, is performed lazily
+func New(beatInfo beat.Info) *Locker {
+	lockfilePath := paths.Resolve(paths.Data, beatInfo.Beat+".lock")
+	return &Locker{
+		fileLock:  flock.New(lockfilePath),
+		logger:    logp.L(),
+		beatName:  beatInfo.Beat,
+		filePath:  lockfilePath,
+		beatStart: beatInfo.StartTime,
+	}
+}
+
+// Lock attempts to acquire a lock on the data path for the currently-running
+// Beat instance. If another Beats instance already has a lock on the same data path
+// an ErrAlreadyLocked error is returned.
+// This lock is pid-aware, and will attempt to recover if the lockfile is taken by a PID that no longer exists.
+func (lock *Locker) Lock() error {
+	// create the pid file that will be used as the lock
+	noFile, err := lock.createPidfile(os.Getpid())
+	if err != nil {
+		return fmt.Errorf("error creating pidfile for lock: %w", err)
+	}
+	if !noFile {
+		err := lock.handleFailedLock()
+		if err != nil {
+			return fmt.Errorf("lock file already existed, error recovering: %w", err)
+		}
+	}
+
+	// This will be a shared/read lock,
+	// if we need to manage a previous lock from another beat instance,
+	// we'll need to read from the pid file
+	isLocked, err := lock.fileLock.TryRLock()
+	if err != nil {
+		return fmt.Errorf("unable to lock data path: %w", err)
+	}
+
+	// case: lock could not be obtained.
+	if !isLocked {
+		// if we're here, things are probably unrecoverable, as we've previously checked for a lockfile. Exit.
+		return ErrAlreadyLocked
+	}
+
+	return nil
+}
+
+// Unlock attempts to release the lock on a data path previously acquired via Lock().
+func (lock *Locker) Unlock() error {
+	err := lock.fileLock.Unlock()
+	if err != nil {
+		return fmt.Errorf("unable to unlock data path: %w", err)
+	}
+
+	err = os.Remove(lock.fileLock.Path())
+	if err != nil {
+		return fmt.Errorf("unable to unlock data path file %s: %w", lock.fileLock.Path(), err)
+	}
+	return nil
+}
+
+// ******* private helpers
+
+// createPidfile creates the underling pidfile used by the lock
+// This assumes that no previous lockfile should exist.
+// returns a bool indicating if the file was created, and an error
+func (lock *Locker) createPidfile(pid int) (bool, error) {
+	_, err := os.Stat(lock.filePath)
```

Comment:
I think there is a data race here.  You check for existance, then write, then get lock.  Between existence check and write someone else could make the file.  It would probably be better to follow the pattern used by golang [CreateTemp](https://cs.opensource.google/go/go/+/refs/tags/go1.19.1:src/os/tempfile.go;l=30;drc=334a591a3f4d868368913328b3e81ddf5b0f46fa)

Also probably a good idea to do the writing when we have the lock.  Locking after the write means someone else could have written to the file before we lock it.

Reply:
Ah, thanks for the tempdir example, good idea.